vamb/__main__.py

```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
+        m = 0
+        for v in d.values():
+            if v:
+                m = max(m, len(str(v).split(";")))
+        return m
+
+    max_levels_true = _max_levels_from_dict(true)
+    max_levels_pred = _max_levels_from_dict(pred)
+    max_levels = max(max_levels_true, max_levels_pred, 1)
+
+    correct_per_level = [0] * max_levels
+    have_truth_per_level = [0] * max_levels
+
+    for c in contigs:
+        pred_levels = _parse_taxonomy_string(pred.get(c), max_levels)
+        true_levels = _parse_taxonomy_string(true.get(c), max_levels)
+
+        for i in range(max_levels):
+            t = true_levels[i]
+            if t is None:
+                continue
+            have_truth_per_level[i] += 1
+            if pred_levels[i] == t:
+                correct_per_level[i] += 1
+
+    levels = [f"Level_{i}" for i in range(max_levels)]
+    accuracy_per_level = [
+        correct_per_level[i] / n_contigs if n_contigs else 0.0
+        for i in range(max_levels)
+    ]
+
+    headers = [
+        "Level",
+        "Correct",
+        "Have_truth",
+        "N_contigs",
+        "Accuracy",
+    ]
+
+    rows = []
+    for i in range(max_levels):
+        rows.append(
+            [
+                levels[i],
+                correct_per_level[i],
+                have_truth_per_level[i],
+                n_contigs,
+                f"{accuracy_per_level[i]:.6f}",
+            ]
+        )
+
+    with open(output_file, "w", newline="") as f:
+        w = csv.writer(f, delimiter="\t")
+        w.writerow(headers)
+        w.writerows(rows)
+
+
+def cross_validate_taxonomy(
+    comp_metadata: vamb.parsecontigs.CompositionMetaData,
+    abundance: np.ndarray,
+    tnfs: np.ndarray,
+    lengths: np.ndarray,
+    out_dir: Path,
+    taxonomy_options: TaxometerOptions,
+    cuda: bool,
+) -> vamb.taxonomy.PredictedTaxonomy:
+    logger.info("Running cross validation for the taxonomy")
+
+    tax = taxonomy_options.taxonomy
+    t = vamb.taxonomy.Taxonomy.from_file(tax.path, comp_metadata, False)
+    n_contigs = len(t.contig_taxonomies)
+    orig_path = str(tax.path)
+    predicted_path = out_dir.joinpath(f"results_taxonomy_predicted_kfold.tsv")
+    file_tracking = out_dir.joinpath(f"file_tracking.tsv")
+    accuracy_file = out_dir.joinpath(f"accuracy_report.tsv")
+
+    all_predicted_taxonomies = []
+    k = 5
+    kf = KFold(n_splits=k, shuffle=True)
+
+    nodes, ind_nodes, table_parent = vamb.taxvamb_encode.make_graph(t.contig_taxonomies)
+    classes_order: list[str] = []
+    for i in t.contig_taxonomies:
+        if i is None or len(i.ranks) == 0:
+            classes_order.append("root")
+        else:
+            classes_order.append(i.ranks[-1])
+    targets = np.array([ind_nodes[i] for i in classes_order])
+
+    def train_and_predict_fold(fold, train_idx, test_idx):
+        logger.info(
+            f"Fold {fold + 1}/{k}: Training on {len(train_idx)} contigs, testing on {len(test_idx)} contigs"
```

Comment:
I'm not a big fan of nested functions - they break up the original function, and the way they read variables from the surrounding scope can be confusing. Move to global scope.